### PR TITLE
[FIX] base: fix batch report outlines traceback

### PR DESCRIPTION
**Pull request by stcc-odoo**
Steps to reproduce:

Accounting app > Customers > Invoices
- Create invoice with customer and a couple of products
- In the terms & conditions tab at the bottom, add 20-30 empty lines
and then a heading (/h1-3) with some text. This insures that the printed
report will have more than 1 page.
- Go back to the Invoices list view
- Select the invoice + another random one > Print > Invoices

A traceback appears in _post_pdf: Assertion error
`assert len(outlines_pages) == len(res_ids)`

When printing a PDF file, `wkhtmltopdf` generates a number of `outlines`
equal to the number of headings (`<h1-6>`) in the printed document.
When printing multiple documents, the logic requires that each document
should have all of its headings on one page.

The traceback is raised because one document has headings on more
than one page.

The information about which `outlines` correspond to which document is
lost during the pdf creation process, so there is no easy way to reason
about the length of each document after the pdf is created.

Solution:

If there are more outlines then expected, then the matching fails and
each part of the PDF will not be saved as attachment, similarly to https://github.com/odoo/odoo/pull/25496

opw-2882507

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
current fix would look good to me

I think we only need to replace H1 tag in general in all our reports (since in all our reports I think we have at least one H1). Replacing also H2 and H6 would have the advantage that a custom report with H2-H6 as top level would be fixed too, but the drawback that possible custom CSS would break would be higher (with H1 you would just break custom css over H1 (that makes no sense in our use case), with H1 to H6 you would break custom CSS targetting H2 to H6.

removing the t-field will remove the heuristic there is here I think:

https://github.com/odoo/odoo/blob/1ad71876d3705f93f51d4b0842fea2d8c8f8fb60/odoo/addons/base/models/ir_qweb_fields.py#L335-L352

maybe we should keep using the widget with something like `t-options="{'widget': 'html'}` if `t-field` doesn't work. But as far as I can tell this doesn't really matter since this is only used by website which is not our case (but this could be used by custom code or in the future).

I think the current explicit solution (I mean "modifying the report template") is the best solution for stable since this is only a concerns for:

- multiple records printed at same time
- that are of a report saved as attachment
- where someone would put Html field containing heading tag of the same level as the report top-level heading tag

other more automatic solution would break more possible custom CSS that would target H1..H6 than necessary. Here we only break it for people that update the module of the report, and only for the affected cases.

in any case, I think there should be a review probably from ORM  (maybe chm or xmo maybe)

**edit:**

note 1: if done, this might be better to target 14.0 instead (if yes, you can keep this PR and rebase on 14.0 and change the target), I think the code is probably a lot similar and it's better to have same code (not 13.0 that will be unsupported soon).

note 2: commit should be prefixed by "account, base", especially since the most important change is in base

note 3: if this is accepted, it would be good to have a small report test that just check this is working

File: odoo/addons/base/models/ir_actions_report.py [hunk 1]
```diff
@@ -963,6 +963,8 @@ def _get_rendering_context(self, docids, data):
                 'docs': docs,
             })
         data['is_html_empty'] = is_html_empty
+        # Replaces unescaped(!) headings (h[1-6]) with <p class='h[1-6]'> tags.
+        data['replace_headings'] = lambda x: Markup(re.sub(r"<h(\d)>(.+?)<\/h\1>", "<p class='h\\1'>\\2</p>", str(x), re.M))
```
Comment (nle-odoo) [inline odoo/addons/base/models/ir_actions_report.py:967]: should be `[1-6]` instead of `\d` if we want H1-H6, there is not much difference

should be insensitive to case (`re.I` flag) even if the editor will have it in lower case

this doesn't allow attribute on h1 (maybe there is not in the editor but we could imagine content in a HTML not from the editor)

we can put a P tag in a H1-H7 tag, but we can't put a P tag in a P tag (eg. browser renderer will correct `<p id="hello"><p>world</p></p>` to `<p id="hello"></p><p>world</p><p></p>`), so if we only fixed H1 maybe a better solution would be to replace H1 by H2. If doing that you could do a simpler regex, only targeting `<h(/?)1\b`, this could even just be a replace without regex (with 2 replace for lower case or 4 with upper case tags).

the regex usually are often compiled one time in the top of the file and referenced (instead of compiled at each function call), but it doesn't seem too important here.

**Comment by nle-odoo** [comment]
as discussed, only targetting H1 doen't work (for example in the main culprit account.invoice report, the top outline is H2)

so I think solution like current one (regex that replace h1-h6 (or at least h1-h2 if we were more restrictive) with classes, would be the good solution (since otherwise the proposed solution would need to transform h1 and h2 to h3)

**Comment by xmo-odoo** [comment]
1. Current version is not great because it can launder arbitrary strings, it should only re-wrap the output if the input was a `Markup`.
2. It's a bit of a variation on @nle-odoo's original suggestion, but I think a good alternative would be to make this an option of the HTML converter. Since the HTML field is already processing every element, it could be done there by swapping out the element's `tag` and adding / setting the relevant class(es).

**Comment by stcc-odoo** [comment]
@xmo-odoo updated

**Comment by odony** [comment]
I have some questions about this PR, please:

- in the PR description I don't understand what "outlines" is and what it has to do with the patch or why it would cause tracebacks? Is this a PDF concept? A qweb concept? An HTML concept?
- I'd like to see the traceback or at least the exception mentioned, or something specific about the actual crash and its cause. I appreciate the reproducer steps and the attempt at explaining the root cause in many words, but it ends up being very unclear technically to me. 
- what is a "faulty field"?
- why does the commit talk about "unescaped things", and what does it have to do with the problem? (if nothing, then please remove it, it's confusing)
- what is `replace_heading` supposed to mean? how will devs know what it does and when to use it in the future?
- looks like you're substituting `<h1>` with `<p class="h1">` or something like that? This kind of example may be useful in the commit msg *and* in a code comment to clarify what you're doing. When the code is doing non-trivial things, or doing things for non-trivial reasons, comments are very important.

In a nutshell, this PR is quite obscure, despite having a lot of words, and the patch is definitely weird 🤨

Moreover, it's modifying a central piece of our qweb engine for something that looks like a rare corner case. That's not a good sign. Are you really sure the problem can't be worked around with a less invasive change?
If it's a rare corner case, it's okay if the solution isn't quite generic or with a broad spectrum, or requires manual steps for the few users who encounter it.

**Comment by stcc-odoo** [comment]
Hi @odony. Indeed the commit message was poorly explained. I hope now it's more clear.

As for the generality of the fix, a less invasive fix was proposed in commit 67c0fddda47b03203fd7babd8f8c07ffc1a9d8d7.
However, as @xmo-odoo pointed out earlier, this poses a security risk as `it can launder arbitrary strings`.

@xmo-odoo, do you think we could instead use your first suggestion like so?
```python
        data['is_html_empty'] = is_html_empty
        # Replaces headings (h[1-6]) with <p class='h[1-6]'> tags.

        def replace_headings(x):
            if type(x) == Markup:
                return Markup(re.sub(r"<h([1-6])>(.+?)<\/h\1>", "<p class='h\\1'>\\2</p>", str(x), re.M | re.I))
            return x
        data['replace_headings'] = replace_headings
        return data
```

File: odoo/addons/base/models/ir_qweb_fields.py [hunk 2]
```diff
@@ -343,7 +343,17 @@ def value_to_html(self, value, options):
...
         for element in body.iter():
+            if replace_headings:
+                if element.tag in ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']:
+                    new_class = element.tag
+                    element.tag = 'p'
+                    classes = element.attrib.get('class', '')
+                    if new_class not in classes:
+                        if classes:
+                            classes += ' '
+                        element.attrib.update({'class': classes + new_class})
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_qweb_fields.py:356]:
```suggestion
                    classes = element.get('class', '').split()
                    if new_class not in classes:
                        element.set('class', ' '.join(classes + [new_class]))
```

**Comment by xmo-odoo** [comment]
> do you think we could instead use your first suggestion like so?

Sure :shrug: though you should probably use `isinstance` rather than `type(x)`, that form is reserved when it's absolutely necessary.

Also `type(x)` should always be used with `is`.

**Comment by odony** [comment]
> Hi @odony. Indeed the commit message was poorly explained. I hope now it's more clear.

Thanks, it's a lot better now!
And now that I understand the issue, I'd like to propose a different resolution path.

The problem stems from our brittle heuristics for "batch printing a single PDF then splitting by outlines to re-attach the pages to individual records".

The thing is, this can simply not work reliably, because there are infinite ways for the logic to be broken and end up with too many or too few outlines: customization of reports, third-party modules, html field values with creative contents, etc.
Yet we're accumulating hacks in order to support all cases. That doesn't seem like a smart strategy.

Also, we have to consider that this class of problem occurs rarely, and only when 3 unusual things occur simultaneously:
- 1) the report is marked to be saved as attachment
- 2) multiple target docs are printed in batch
- 3) there is something specific that causes extra outline levels to be produced (that's the part where infinite problems can happen)

Moreover, users don't usually batch-print documents in order to trigger the auto-save-as-attachment feature. They do it because they need to print multiple PDFs at once.

Based on this, **why don't we instead make the framework tolerant to these cases**, and simply ignore the "save as attachment" part when the outline matching heuristics fail?
Either change the logic to detect the fail conditions and skip, or wrap the whole thing in a try/except, with a technical log of the issue and a fallback strategy.

In fact that's somewhat the strategy @nle-odoo used in the past when there was zero outline, cfr #25496.

IMO this leads to the same result: no crash for customers, except with a much better chance of success of stopping many more variations of this error, without requiring an endless pile of hacks.

File: odoo/addons/base/models/ir_actions_report.py [hunk 3]
```diff
@@ -741,27 +741,32 @@ def close_streams(streams):
...
-                            attachment_writer.write(stream)
-                            if res_ids[i] and res_ids[i] not in save_in_attachment:
-                                new_stream = self._postprocess_pdf_report(record_map[res_ids[i]], stream)
-                                # If the buffer has been modified, mark the old buffer to be closed as well.
-                                if new_stream and new_stream != stream:
-                                    close_streams([stream])
-                                    stream = new_stream
-                            streams.append(stream)
-                        close_streams([pdf_content_stream])
+                        if len(outlines_pages) == len(res_ids):
```
Comment (nle-odoo) [inline odoo/addons/base/models/ir_actions_report.py:744]: maybe we could have:

```py
outlines_pages = []
if '/Outlines' in root and '/First' in root['/Outlines']:
    outlines_pages = []
    node = root['/Outlines']['/First']
    while True:
        outlines_pages.append(root['/Dests'][node['/Dest']][0])
        if '/Next' not in node:
            break
        node = node['/Next']
    outlines_pages = sorted(set(outlines_pages))
if len(outlines_pages) == len(res_ids) and outlines_pages[0] == 0:
```
and not duplicate the else (in this branch `len(res_ids) > 1`)

**Review by nle-odoo — APPROVED** [review]
new solution looks good to me :+1: 

there is still one assertion:

```py
# There should be a top-level heading on first page
assert outlines_pages[0] == 0
```
that would fail, and I think now we should also handle it similarly.

File: odoo/addons/base/models/ir_actions_report.py [hunk 4]
```diff
@@ -741,27 +741,32 @@ def close_streams(streams):
                                 break
                             node = node['/Next']
```
Comment (nle-odoo) [inline odoo/addons/base/models/ir_actions_report.py:742]: maybe we could have:

```py
outlines_pages = []
if '/Outlines' in root and '/First' in root['/Outlines']:
    node = root['/Outlines']['/First']
    while True:
        outlines_pages.append(root['/Dests'][node['/Dest']][0])
        if '/Next' not in node:
            break
        node = node['/Next']
    outlines_pages = sorted(set(outlines_pages))
if len(outlines_pages) == len(res_ids) and outlines_pages[0] == 0:
```
and not duplicate the else (in this branch `len(res_ids) > 1`)

File: odoo/addons/base/models/ir_actions_report.py (hunk 3 above)
Comment (nle-odoo) [inline odoo/addons/base/models/ir_actions_report.py:744]: maybe we could have:

```py
outlines_pages = []
if '/Outlines' in root and '/First' in root['/Outlines']:
    node = root['/Outlines']['/First']
    while True:
        outlines_pages.append(root['/Dests'][node['/Dest']][0])
        if '/Next' not in node:
            break
        node = node['/Next']
    outlines_pages = sorted(set(outlines_pages))
if len(outlines_pages) == len(res_ids) and outlines_pages[0] == 0:
```
and not duplicate the else (in this branch `len(res_ids) > 1`)

**Comment by stcc-odoo** [comment]
@odony Does it look good to you?

**Comment by stcc-odoo** [comment]
@odony up

**Review by nle-odoo — APPROVED** [review]
looks good to me, now we should have:

|case | before | after |
|-|-|-|
|no outline | no save, no error | no save, no error, info log |
|more outlines than pages | no save, assert error | no save, no error, info log |
|first outline not on first page | no save, assert error | no save, no error, info log |
|correct number of outlines | attachment save | attachment save |

**Comment by stcc-odoo** [comment]
@odony Could you please review the changes?

**Review by odony — APPROVED** [review]
That looks very nice and much more resilient/future-proof, thanks! :+1:

**Comment by odony** [comment]
@robodoo r+
